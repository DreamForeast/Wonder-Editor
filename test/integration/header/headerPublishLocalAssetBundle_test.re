open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

open Js.Promise;

open Js.Typed_array;

let _ =
  describe("header publish local->asset bundle", () => {
    let sandbox = getSandboxDefaultVal();

    beforeEach(() => {
      sandbox := createSandbox();

      MainEditorSceneTool.initState(~sandbox, ());

      MainEditorSceneTool.prepareScene(sandbox);

      MainEditorAssetTool.buildFakeFileReader();
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test ui", () =>
      testPromise("test show select tree with only asset bundles", () => {
        let addedMaterialNodeId1 = MainEditorAssetIdTool.getNewAssetId();
        MainEditorAssetHeaderOperateNodeTool.addMaterial();

        let addedFolderNodeId1 = MainEditorAssetIdTool.getNewAssetId();
        MainEditorAssetHeaderOperateNodeTool.addFolder();

        MainEditorAssetTreeTool.Select.selectFolderNode(
          ~nodeId=addedFolderNodeId1,
          (),
        );

        MainEditorAssetUploadTool.loadOneAssetBundle(~fileName="A.wab", ())
        |> then_(uploadedAssetBundleNodeId1 =>
             MainEditorAssetUploadTool.loadOneAssetBundle(
               ~fileName="B.sab",
               (),
             )
             |> then_(uploadedAssetBundleNodeId2 =>
                  HeaderPublishLocalAssetBundleTool.buildPublishLocalModal(
                    ~defaultUseAssetBundle=true,
                    (),
                  )
                  |> BuildComponentTool.buildUI
                  |> ReactTestTool.createSnapshotAndMatch
                  |> resolve
                )
           );
      })
    );

    describe("download zip", () =>
      describe("test use assetbundle", () => {
        let _prepare =
            (~judgeFunc, ~selectTreeForAssetBundle, ~useAssetBundle=true, ()) => {
          let fakeFetchFunc = PublishLocalTool.buildFakeFetch(~sandbox, ());

          let obj = HeaderTool.buildPublishFakeJsZipCreateFunc(sandbox^);

          HeaderPublishLocalUtils.Publish.publishZip(
            ("WonderLocal", false),
            (useAssetBundle, selectTreeForAssetBundle),
            () => obj,
            fakeFetchFunc,
          )
          |> then_(_ => {
               let file = obj##file;
               let fetchCount =
                 PublishLocalTool.getFetchPackageContentWithoutAssetCountWithDefault();

               judgeFunc(fetchCount, file) |> resolve;
             });
        };

        describe("export selected asset bundles", () => {
          beforeEach(() => {
            LoadTool.buildFakeTextEncoder();

            LoadTool.buildFakeURL(sandbox^);

            LoadTool.buildFakeLoadImage(.);
          });

          testPromise("test", () => {
            let addedMaterialNodeId1 = MainEditorAssetIdTool.getNewAssetId();
            MainEditorAssetHeaderOperateNodeTool.addMaterial();

            let addedFolderNodeId1 = MainEditorAssetIdTool.getNewAssetId();
            MainEditorAssetHeaderOperateNodeTool.addFolder();

            MainEditorAssetTreeTool.Select.selectFolderNode(
              ~nodeId=addedFolderNodeId1,
              (),
            );

            MainEditorAssetUploadTool.loadOneAssetBundle(
              ~fileName="A.wab",
              (),
            )
            |> then_(uploadedAssetBundleNodeId1 =>
                 MainEditorAssetUploadTool.loadOneAssetBundle(
                   ~fileName="B.sab",
                   (),
                 )
                 |> then_(uploadedAssetBundleNodeId2 => {
                      let selectTree =
                        HeaderPublishLocalAssetBundleTool.buildSelectTreeForAssetBundle
                        |> StateLogicService.getStateToGetData
                        |> SelectTreeTool.setSelectForSelectTree(
                             true,
                             MainEditorAssetAssetBundleNodeTool.getName(
                               uploadedAssetBundleNodeId2,
                             )
                             |> StateLogicService.getEditorState,
                           );

                      _prepare(
                        ~selectTreeForAssetBundle=selectTree,
                        ~judgeFunc=
                          (fetchCount, file) =>
                            file
                            |> getCall(0)
                            |> getArgs
                            |> Js.List.hd
                            |> OptionService.unsafeGet
                            |> expect == "AssetBundles/New Folder/B.sab",
                        (),
                      );
                    })
               );
          });
        });
      })
    );
  });