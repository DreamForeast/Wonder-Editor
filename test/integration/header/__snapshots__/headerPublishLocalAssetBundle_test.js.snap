// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header publish local->asset bundle test ui test show select tree with only asset bundles 1`] = `
<article
  className="wonder-singleInput-modal"
>
  <div
    className="modal-item"
  >
    <div
      className="modal-item-header"
    >
      publish local
      <img
        onClick={[Function]}
        src="./public/img/close.png"
      />
    </div>
    <div
      className="modal-item-content"
    >
      <div
        className="content-field"
      >
        <div
          className="field-title"
        >
          name
        </div>
        <div
          className="field-content"
        >
          <input
            className="input-component"
            onChange={[Function]}
            type="text"
            value="WonderLocal"
          />
        </div>
      </div>
      <div
        className="content-field"
      >
        <div
          className="field-title"
        >
          useWorker
        </div>
        <div
          className="field-content"
        >
          <input
            defaultChecked={false}
            onClick={[Function]}
            type="checkbox"
          />
        </div>
      </div>
      <div
        className="content-field"
      >
        <div
          className="field-title"
        >
          useAssetBundle
        </div>
        <div
          className="field-content"
        >
          <input
            defaultChecked={true}
            onClick={[Function]}
            type="checkbox"
          />
        </div>
      </div>
      <article
        className="wonder-selectTree"
      >
        <ul
          className="wonder-tree-node"
        >
          <li>
            <div
              className="item-triangle"
            >
              <img
                src="./public/img/down.png"
              />
            </div>
            <div
              className="select-box"
            >
              <input
                defaultChecked={false}
                onClick={[Function]}
                type="checkbox"
              />
            </div>
            <img
              className="treeNode-icon"
              src="./public/img/package.png"
            />
            <span>
              Assets
            </span>
          </li>
          <ul
            className="wonder-tree-node"
          >
            <li>
              <div
                className="item-triangle"
              >
                <img
                  src="./public/img/down.png"
                />
              </div>
              <div
                className="select-box"
              >
                <input
                  defaultChecked={false}
                  onClick={[Function]}
                  type="checkbox"
                />
              </div>
              <img
                className="treeNode-icon"
                src="./public/img/package.png"
              />
              <span>
                New Folder
              </span>
            </li>
            <ul
              className="wonder-tree-node"
            >
              <li>
                <div
                  className="item-triangle"
                />
                <div
                  className="select-box"
                >
                  <input
                    defaultChecked={false}
                    onClick={[Function]}
                    type="checkbox"
                  />
                </div>
                <img
                  className="treeNode-icon"
                  src="./public/img/assetBundle.png"
                />
                <span>
                  A
                </span>
              </li>
            </ul>
            <ul
              className="wonder-tree-node"
            >
              <li>
                <div
                  className="item-triangle"
                />
                <div
                  className="select-box"
                >
                  <input
                    defaultChecked={false}
                    onClick={[Function]}
                    type="checkbox"
                  />
                </div>
                <img
                  className="treeNode-icon"
                  src="./public/img/assetBundle.png"
                />
                <span>
                  B
                </span>
              </li>
            </ul>
          </ul>
        </ul>
      </article>
    </div>
    <div
      className="modal-item-footer"
    >
      <button
        className="footer-submit"
        onClick={[Function]}
      >
        Submit
      </button>
    </div>
  </div>
</article>
`;
